rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+
+##### Example notebook usage
+
+Here is an example of what notebook execution may look like in this scheme.
+
+```
+Input[0]:
+# To begin, we initialize an interactive context. Here, by not passing
+# in a base directory or metadata configuration, we create an ephemeral
+# context whose outputs will be in a temporary directory.
+context = InteractiveContext()
+
+# Alternatively, we may pass in these directories for a context using a
+# persistent store:
+#
+# context = InteractiveContext(base_dir=my_base_dir,
+#                              metadata_connection_config=my_config)
+```
+
+<table>
+  <tr>
+   <td><strong>Input[1]:</strong>
+<p>
+<code><em># First, ExampleGen with a run / read.</em></code>
+<code>example_gen = CsvExampleGen(input_base=examples)</code>
+<p>
+<code><em># Note that the output will be of type 'ExamplesPath', for which we</em></code>
+<code><em># may have registered a notebook visualization handler.</em></code>
+<code>example_gen_result = context.run(example_gen)</code>
+<p>
+<code>example_gen.outputs['examples']</code>
+<p>
+<code><em># alternative style: explicit context.show() method</em></code>
+<code>context.show(example_gen.outputs['examples'])</code>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>Output[1]:</strong>
+<p>
+_notebook visualization indicating we have N examples at some temp path_
+   </td>
+  </tr>
+</table>
+
+<table>
+  <tr>
+   <td><strong>Input[2]:</strong>
+<p>
+<code><em># Next, StatisticsGen with a run / read.</em></code>
+<code>statistics_gen = StatisticsGen(input_data=example_gen.outputs['examples'])</code>
+<p>
+<code>context.run(statistics_gen).outputs['output']</code>
+<p>
+<code><em># alternative styles:</em></code>
+<code><em># context.<strong>show(</strong>context.run(statistics_gen).outputs['output'])</em></code>
+<code><em># context.run().read('output'<strong>, visualization_handler=blah</strong>)</em></code>
+<code><em># context.run().<strong>show</strong>('output', <strong>visualization_handler=blah, visualization_args=</strong>)</em></code>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>Output[2]:</strong>
+<p>
+_notebook visualization for statsgen output_
+   </td>
+  </tr>
+</table>
+
+<table>
+  <tr>
+   <td><strong>Input[3]:</strong>
+<p>
+<code><em># Next, SchemaGen without a run / read.</em></code>
+<code>infer_schema = SchemaGen(stats=statistics_gen.outputs['output'])</code>
+<p>
+<code><em># Finally, ExampleValidator with a run / read. Note that SchemaGen</em></code>
+<code><em># will be implicitly run (see Note 2 below).</em></code>
+<code>validate_stats = ExampleValidator(</code>
+<p>
+<code>    stats=statistics_gen.outputs['output'],</code>
+<p>
+<code>    schema=infer_schema.outputs['output'])</code>
+<p>
+<code>context.run(validate_stats)</code>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>Output[3]:</strong>
+<p>
+_ExecutionResult object for ExampleValidator_
+   </td>
+  </tr>
+</table>
+
+Note that the user may have forgotten to run InteractiveContext.run() on
```

Comment:
I had the same question as Alex, and the answer has been given in the same way a couple times now, which makes me think: maybe the problem here is that we are trying to "translate" the existing pipeline runner to colab, and Alex and I, as uninitiated users, are confused as to why this isn't more natively colab-like. Specifically, by its nature, a DAG is hard to handle in colab, since the user's expectation is that cells can be run independently to some extent. Plus, colab tends to assume everything is transient, whereas the pipeline relies on artifacts.

Which brings me back to the question at the top-- is the target user an existing TFX user who wants flexibility in notebooks, or is this designed to capture new users who want ML pipelines from notebooks? If the latter, we may want to consider a more notebook-first paradigm, rather than retrofitting the existing pipeline to colab. If, however, the expectation is that the primary use-case will still be the normal pipeline, it's fair to assume the user will end up being familiar with the DAG and prefer uniformity here.

Reply:
We want to target users who want to build TFX pipelines iteratively, so the end goal is that the user is taking what they built and running it on an orchestrator (e.g. Beam, Airflow or Kubeflow) outside of the notebook.

So while we do want to capture users who want to build ML pipelines, we want to do so in a way that (1) does not diverge too much from how such pipelines would be configured in "production TFX" and (2) will expose them gradually to TFX concepts like the component DAG (starting out by running component cells individually to produce outputs, and building the DAG from components that reference previously generated outputs).